### PR TITLE
[Serve] Bump controller concurrency limit and flag off long poll timeout

### LINKED ISSUE #19704: [nightly tests] [serve] multi_deployment_1k_noop_replica nightly test failing consistently

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?

This should hopefully fix the multi-deployment test. The symptom i'm see so far is basically controller is still too overloaded because multi-deployment has 100 handles per replicas, this amounts to about 100*1000 long poll clients atm (which is pretty bad by itself, I bumped the priority on #19157). 

The fix here is to introduce client side sleep to let controller catch up on other requests. 

## Related issue number

Closes #19704

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
I will verify it with test wheels, but the code is ready for review.

File: python/ray/serve/long_poll.py [hunk 1]
```diff
@@ -17,7 +18,13 @@
...
 # We randomly select a timeout within this range to avoid a "thundering herd"
 # when there are many clients subscribing at the same time.
-LISTEN_FOR_CHANGE_REQUEST_TIMEOUT_S = (30, 60)
+LISTEN_FOR_CHANGE_REQUEST_TIMEOUT_S = (10, 30)
+# Same mechanism as the LISTEN_FOR_CHANGE_REQUEST_TIMEOUT_S but this should
+# be used at client side to sleep for a duration to let long poll host catch
+# up to other requests. If the host returned timeout, this means there is
+# unlikely to be any meaninful changes, so it is ok to wait a bit before
+# pulling again.
+CLIENT_SIDE_TIMEOUT_WAIT_S = (10, 60)
```
Comment (edoakes) [inline python/ray/serve/long_poll.py:27]: this is a really long delay -- wouldn't this mean that replica updates may be delayed by up to a minute? that does not seem acceptable to me

Comment (simon-mo) [reply python/ray/serve/long_poll.py:27]: ideally we want to be longer than the long poll timeout. and the thing here is that if we are already timing out, it is very likely we will keep timing out. 

how about (5,30)?

**Comment by simon-mo** [comment]
@edoakes multi_deploy now passed! https://buildkite.com/ray-project/periodic-ci/builds/1639#cc4a7d43-3c14-456b-8a71-d930489b52b4 PTAL at the PR again

File: python/ray/serve/constants.py [hunk 2]
```diff
@@ -61,4 +71,8 @@
 
 #: Because ServeController will accept one long poll request per handle, its
 #: concurrency needs to scale as O(num_handles)
-CONTROLLER_MAX_CONCURRENCY = 15000
+CONTROLLER_MAX_CONCURRENCY = int(
+    _getenv_default(
+        "SERVE_CONTROLLER_MAX_CONCURRENCY",
+        "1000000",
```
Comment (edoakes) [inline python/ray/serve/constants.py:77]: comment why this number

**Review by edoakes — APPROVED** [review]

**Comment by simon-mo** [comment]
image
gcs_server seems to struggling with node launch failure. but let me try again.

File: release/serve_tests/workloads/multi_deployment_1k_noop_replica.py [hunk 3]
```diff
@@ -143,13 +143,11 @@ def main(num_replicas: Optional[int], num_deployments: Optional[int],
     setup_multi_deployment_replicas(num_replicas, num_deployments)
 
     logger.info("Warming up cluster ....\n")
-    rst_ray_refs = []
     all_endpoints = list(serve.list_deployments().keys())
     for endpoint in all_endpoints:
-        rst_ray_refs.append(
+        ray.get(
```
Comment (jiaodong) [inline release/serve_tests/workloads/multi_deployment_1k_noop_replica.py:148]: is this an attempt to limit concurrent cluster warm up by doing it one at a time ?

Comment (simon-mo) [reply release/serve_tests/workloads/multi_deployment_1k_noop_replica.py:148]: yup

**Comment by simon-mo** [comment]
from offline discussion:

Here is my understanding:
- We can increase the max_concurrency for controller. Increasing it too high will cause Python stack to freak out due to ray asyncio implementation.
- Just max_concurrency doesn’t work, because we have 100*1000 long poll requests, each of them will timeout between 30->60s.  This means we need controller to handle (100000/45 ~= 2k) new long poll request per second. This is very much hitting the limit of ray actor calls so controller is spinning 100% cpu.
- We can increase the timeout, but then serve_failure test will become more flaky. Because it will then takes forever for the old long poll connection to clean up and use the max_concurrency slot.
- There are few ways to solve this, both requires more thinking
  - let the client wait for some time before sending a new long poll request
  - let long poll receives the client death notification somehow to automatically clean it up. for long poll coming from serve 
     - replicas that’s easy because we already have healthcheck ref.
  - use push based notification for clients in ray actors, pull based for drivers
  - go ahead and merge all long poll request to single client so it will be global long poll client per process.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.
